### PR TITLE
Fixed sampling

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:
Fixed sampling implemented -- samples across epochs can stay the same

In some cases, people want to have the same samples every epoch. This keeps the seed between epochs the same, enabling this behavior.

## Issue #, if available:
Addresses STR-36

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [x] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -313,6 +327,22 @@ def __init__(self,
             }
             for stream in streams:
                 stream.apply_default(default)
+        elif epoch_size_value:
```
Comment (karan6181) [inline streaming/base/dataset.py:330]: I think you might want to revert this to the older PR approach where you included this logic in below `else` section !!!

File: tests/test_streaming.py [hunk 2]
```diff
@@ -59,6 +59,65 @@ def test_dataloader_epoch_size_no_streams(local_remote_dir: Tuple[str,
...
+                               sampling_method='fixed')
+
+    # Build DataLoader
+    dataloader = StreamingDataLoader(dataset=dataset,
+                                     batch_size=batch_size,
+                                     num_workers=num_workers,
+                                     drop_last=drop_last)
+
+    # iterate once over the dataloader (first epoch)
+    first_samples_seen = {}
```
Comment (karan6181) [inline tests/test_streaming.py:96]: Can you combine below two for loops into one ?

File: tests/test_streaming.py [hunk 3]
```diff
@@ -59,6 +59,65 @@ def test_dataloader_epoch_size_no_streams(local_remote_dir: Tuple[str,
...
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(local=local_dir,
+                               remote=remote_dir,
+                               shuffle=shuffle,
+                               batch_size=batch_size,
+                               shuffle_seed=seed,
+                               num_canonical_nodes=num_canonical_nodes,
+                               epoch_size=epoch_size,
+                               sampling_method='fixed')
```
Comment (karan6181) [inline tests/test_streaming.py:87]: Can you test it for both the sampling method ? You can `@pytest.mark.parametrize` for the same and have `if` loop logic below to check if the sample order is deterministic for `fixed` sampling method and not for `balanced` sampling method !!

File: streaming/base/dataset.py [hunk 4]
```diff
@@ -284,12 +292,18 @@ def __init__(self,
...
         self.shuffle_block_size = shuffle_block_size
+        self.sampling_method = sampling_method.lower().strip()
 
         # Check streams vs remote/local.
         if bool(streams) == (bool(remote) or bool(local)):
             raise ValueError(
                 'You must provide either `streams` or `remote`/`local`, but not both.')
 
+        # Check sampling method is one of "balanced" or "fixed".
+        if self.sampling_method not in ['balanced', 'fixed']:
```
Comment (karan6181) [inline streaming/base/dataset.py:303]: Can you also add a test case for this ?

File: streaming/base/dataset.py (hunk 1 above)
Comment (snarayan21) [reply streaming/base/dataset.py:330]: yep my mistake. bad merge

File: tests/test_streaming.py (hunk 2 above)
Comment (snarayan21) [reply tests/test_streaming.py:96]: bad merge again...

File: streaming/base/dataset.py (hunk 4 above)
Comment (snarayan21) [reply streaming/base/dataset.py:303]: yep!

File: tests/test_streaming.py (hunk 3 above)
Comment (snarayan21) [reply tests/test_streaming.py:87]: fixed 👍

File: tests/test_streaming.py [hunk 5]
```diff
@@ -181,6 +175,30 @@ def test_dataloader_single_device(local_remote_dir: Tuple[str, str], batch_size:
         assert len(set(sample_order)) == num_samples
 
 
+@pytest.mark.parametrize('batch_size', [1, 4])
```
Comment (karan6181) [inline tests/test_streaming.py:178]: Minor comment: Since we are just testing exception, we can fixed the other parameters which is not directly impacting the exception such as `batch_size`, `shuffle`, `num_workers` etc. Something like this

```
@pytest.mark.parametrize('batch_size', [4])
@pytest.mark.parametrize('seed', [1111])
@pytest.mark.parametrize('shuffle', [True])
@pytest.mark.parametrize('num_workers', [2])
@pytest.mark.usefixtures('local_remote_dir')
```

Comment (snarayan21) [reply tests/test_streaming.py:178]: Cool sounds good

**Review by karan6181 — APPROVED** [review]
